### PR TITLE
engine: fake watcher sends events to all watches

### PR DESCRIPTION
**Pull request by jazzdan**
Pulled out of #757

### CODE REVIEW

**Comment by jazzdan** [comment]
hmm, I thought this was passing but it looks like it's flaky.

**Comment by jazzdan** [comment]
All good now, review away!

File: internal/engine/fake_notify_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	defer w.mu.Unlock()
+	w.subs = append(w.subs, subCh)
+	w.subsErrors = append(w.subsErrors, errorCh)
+	return newFakeWatcher(subCh, errorCh), nil
+}
+
+func (w *fakeMetaWatcher) getSubs() []chan watch.FileEvent {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.subs
```
Comment (nicks) [inline internal/engine/fake_notify_test.go]: this probably should copy the slice?

File: internal/engine/fake_notify_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+func (w *fakeMetaWatcher) getSubErrors() []chan error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.subsErrors
```
Comment (nicks) [inline internal/engine/fake_notify_test.go]: ditto

File: internal/engine/fake_notify_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+func (w *fakeMetaWatcher) loop() {
+	for {
+		select {
+		case e, ok := <-w.events:
+			if !ok {
+				for _, sub := range w.getSubs() {
```
Comment (nicks) [inline internal/engine/fake_notify_test.go]: probably more correct to put this in a defer at the top of the function? bc you want to close this in the other `return` case as well

File: internal/engine/fake_notify_test.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
+package engine
+
+import (
+	"sync"
+
+	"github.com/windmilleng/tilt/internal/watch"
+)
+
+type fakeMetaWatcher struct {
```
Comment (nicks) [inline internal/engine/fake_notify_test.go]: maybe MultiWatcher instead of MetaWatcher?

File: internal/engine/upper_test.go [hunk 5]
```diff
@@ -230,6 +199,7 @@ func TestUpper_UpWatchCoalescedFileChanges(t *testing.T) {
 	for _, fileRelPath := range fileRelPaths {
 		f.fsWatcher.events <- watch.FileEvent{Path: fileRelPath}
 	}
+	time.Sleep(time.Millisecond)
```
Comment (nicks) [inline internal/engine/upper_test.go]: ?

this seems like it means you introduced a bug

File: internal/engine/upper_test.go [hunk 6]
```diff
@@ -1600,6 +1561,7 @@ func (f *testFixture) Start(manifests []model.Manifest, watchMounts bool) {
 	f.WaitUntil("manifests appear", func(st store.EngineState) bool {
 		return len(st.ManifestStates) == len(manifests) && st.WatchMounts == watchMounts
 	})
+	time.Sleep(5 * time.Millisecond)
```
Comment (nicks) [inline internal/engine/upper_test.go]: ditto here, this seems like an obvious bug

File: internal/engine/fake_notify_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	return w.outboundCh
+}
+
+func (w *fakeWatcher) loop() {
+	var q []watch.FileEvent
+	for {
+		var outboundCh chan watch.FileEvent
+		var outboundE watch.FileEvent
+		if len(q) > 0 {
+			outboundCh, outboundE = w.outboundCh, q[0]
```
Comment (nicks) [inline internal/engine/fake_notify_test.go]: this loop is really hard to follow.

it might be a lot simpler if you just have two entirely different branches, one if q is empty and one if it's non-empty. then a lot of the weird outbound variables go away

File: internal/engine/upper_test.go (hunk 5 above)
Comment (jazzdan) [reply internal/engine/upper_test.go]: Yeah what we really need is a bit on the state that says "we have successfully watched the watches." Right now there's a slight delay between when we tell Tilt to start the watches and when they actually start.

I'll make a story for this.

File: internal/engine/fake_notify_test.go (hunk 7 above)
Comment (jazzdan) [reply internal/engine/fake_notify_test.go]: I took another pass, let me know what you think!

File: internal/engine/fake_notify_test.go (hunk 1 above)
Comment (nicks) [reply internal/engine/fake_notify_test.go]: i think you may have misunderstood my comment? i'll just send you a pr

Comment (jazzdan) [reply internal/engine/fake_notify_test.go]: Sounds good!
